### PR TITLE
damldocs: Recognize generic templates, choices, and template instances.

### PR DESCRIPTION
**Pull request by associahedron**
This PR focuses only on recognizing generic templates, choices, and template instances (see #1387), so they don't end up mixed in with other docs. This should be enough to maintain backward compatibility once we switch to the new template desugaring.

This PR does not generate any documentation for template instances, and doesn't add any more documentation for generic templates (e.g. it doesn't document generic template arguments). That's work for another PR.

The main difference in recognition is to allow arguments to templates and choices (by using `hsTyGetAppHead_maybe` to get the name of the type constructor) and allow contexts (by using `getLHsInstDeclHead` to strip out the context). 

Recognizing a template instance is actually hard, given that the generic template that a template instance corresponds to may be in a totally different module, and whether a type is a template is not captured in the typechecked ast without some serious digging. I talked with @hurryabit, and he proposed that we change the name of the instance typeclass (e.g. `ProposalInstance`) to something the user can't generate (e.g. `Proposal:Instance`), so we can recognize it instantly. Another alternative is to include a marker in the docs for a template instance newtype, but this might be a bit more finicky. This PR just assumes that any class ending in `Instance` comes from a generic template ... it's a bit of an over approximation.

Also I took the `ProposalDSL.daml` test file and made a golden test for generic templates out of it. Once generic templates are the standard, we can revert some of the changes I had to make to make this work in the damldoc tests, like adding an import path to the daml-test-files folder or recognizing "Template" and "Choice" instances without the usual qualification ("DA.Internal.Desugar.Template" and "DA.Internal.Desugar.Choice" respectively).

### CODE REVIEW

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs [hunk 1]
```diff
@@ -368,16 +374,20 @@ getTemplateDocs DocCtx{..} typeMap = map mkTemplateDoc $ Set.toList dc_templates
 -- | Extracts all names of Template instances defined in a module and a map of
 -- template to set of its choices (instances of Choice with a particular
 -- template).
-getTemplateData :: ParsedModule -> (Set.Set Typename, MS.Map Typename (Set.Set Typename))
+getTemplateData :: ParsedModule ->
```
Comment (jberthold-da) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs]: `<nit-pick>` haddock description is not correct any more

Comment (associahedron) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs]: that's a good point :-)

File: compiler/damlc/tests/daml-test-files/ProposalDSL.EXPECTED.rst [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+Module ProposalDSL
+------------------
+
+
+Templates
+^^^^^^^^^
+
+.. _template-proposaldsl-proposal-62786:
+
+template **Proposal**
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/ProposalDSL.EXPECTED.rst]: Should we show the type parameter here?

Comment (associahedron) [reply compiler/damlc/tests/daml-test-files/ProposalDSL.EXPECTED.rst]: Yep, but I left that for a separate PR, this one is just for recognizing generic templates (and non-generic templates under the new desugaring).

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs (hunk 1 above)
Comment (associahedron) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs]: Fixed!

File: compiler/damlc/tests/daml-test-files/ProposalDSL.EXPECTED.rst (hunk 2 above)
Comment (cocreature) [reply compiler/damlc/tests/daml-test-files/ProposalDSL.EXPECTED.rst]: :+1:

**Pull request merged**